blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/Token.java

```diff
@@ -0,0 +1,147 @@
+package com.rackspacecloud.blueflood.types;
+
+import org.apache.commons.lang.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ *
+ * For a given tenantId = '111111', metricName = 'a.b.c.d',
+ *
+ * level 2 token will have the following information.
+ *
+ *      token = 'c'
+ *      parent = 'a.b'
+ *      isLeaf = false
+ *      id = '111111:a.b.c'
+ *      tenantId = '111111'
+ *
+ * level 3 token which is a leaf node will have the following information
+ *
+ *      token = 'd'
+ *      parent = 'a.b.c'
+ *      isLeaf = true
+ *      id = '111111:a.b.c.d:$'
+ *      tenantId = '111111'
+ *
+ */
+public class Token {
+
+    public static final String LEAF_NODE_SUFFIX = ":$";
+    public static final String SEPARATOR = ":";
+
+    private final String token;
+    private final String parent;
+    private final boolean isLeaf;
+
+    private final String id;
+    private final Locator locator;
+
+    public Token(Locator locator, String[] tokens, int level) {
```

Comment:
Thats the constructor we use. It has to be public. I will add some documentation.